packages/vue-app/template/client.js

```diff
@@ -146,9 +146,12 @@ async function loadAsyncComponents(to, from, next) {
       // Add a marker on each component that it needs to refresh or not
       const startLoader = Components.some((Component) => {
         const watchQuery = Component.options.watchQuery
-        if (watchQuery === true) return true
-        if (Array.isArray(watchQuery)) {
+        if (watchQuery === true) {
+          return true
+        } else if (Array.isArray(watchQuery)) {
           return watchQuery.some(key => this._diffQuery[key])
+        } else if (typeof watchQuery === 'function') {
+          return watchQuery(from.query, to.query)
```

Comment:
What if calling it with `this` context?

Reply:
The new component hasn’t been created and I think pointing to old instance is not very appropriate

Reply:
We can use it to access general context like store, ...

Reply:
Maybe add it as arg instead `({ from, to, fromComponent })`? Agree with @clarkdo that setting the old component as `this` could be confusing

Reply:
But it makes sense as we are _watching_ from the current component. Like other watchers.

Reply:
Thats also a good point :)

Reply:
Why dismissed?

Reply:
GitHub automatically marked it as outdated? Sorry, I’ll apply change

Reply:
Done, please review